hoc/addons/usr/share/man/man1/hoc.1
.\" -*-nroff-*-
.\" ====================================================================
.\"  @Troff-man-file{
.\"     author          = "Nelson H. F. Beebe",
.\"     version         = "7.0.6.beta",
.\"     date            = "07 August 2002",
.\"     time            = "19:22:44 MDT",
.\"     filename        = "hoc.1",
.\"     address         = "Center for Scientific Computing
.\"                        University of Utah
.\"                        Department of Mathematics, 322 INSCC
.\"                        155 S 1400 E RM 233
.\"                        Salt Lake City, UT 84112-0090
.\"                        USA",
.\"     telephone       = "+1 801 581 5254",
.\"     FAX             = "+1 801 585 1640, +1 801 581 4148",
.\"     URL             = "http://www.math.utah.edu/~beebe",
.\"     checksum        = "25456 5188 19182 126394",
.\"     email           = "beebe@math.utah.edu, beebe@acm.org,
.\"                        beebe@computer.org, beebe@ieee.org
.\"                        (Internet)",
.\"     codetable       = "ISO/ASCII",
.\"     keywords        = "calculator; floating-point arithmetic; hoc
.\"                        programming language; IEEE 754; software
.\"                        internationalization",
.\"     supported       = "yes",
.\"     docstring       = "This file is the UNIX nroff/troff manual page
.\"                        documentation for hoc, the extended
.\"                        high-order calculator.
.\"
.\"                        The checksum field above contains a CRC-16
.\"                        checksum as the first value, followed by the
.\"                        equivalent of the standard UNIX wc (word
.\"                        count) utility output of lines, words, and
.\"                        characters.  This is produced by Robert
.\"                        Solovay's checksum utility.",
.\"  }
.\" ====================================================================
.TH HOC 1 09-Aug-2002 7.06.beta
.\" .CT 1 numbers
.\" ----------------------------------------------
.SH NAME
hoc \(em (high-order calculator) [interactive floating-point language]
.\" ----------------------------------------------
.SH SYNOPSIS
.B hoc
[
.B \-author
]
[
.B \-copyright
]
[
.BI \-D name
]
.if n .ti +\w'\fBhoc\fP 'u
[
.BI \-D name=number
]
[
.BI \-D name="string"
]
.if t .ti +\w'\fBhoc\fP 'u
.if n .ti +\w'\fBhoc\fP 'u
[
.BI \-D name=symbol
]
[
.B \-?
]
[
.B \-help
]
[
.B \-no-banner
]
.if n .ti +\w'\fBhoc\fP 'u
[
.B \-no-cd
]
[
.B \-no-help-file
]
[
.B \-no-load
]
.if n .ti +\w'\fBhoc\fP 'u
.if t .ti +\w'\fBhoc\fP 'u
[
.B \-no-logfile
]
[
.B \-no-readline
]
[
.B \-no-save
]
.if n .ti +\w'\fBhoc\fP 'u
[
.B \-no-site-file
]
[
.B \-no-translation-file
]
.if n .ti +\w'\fBhoc\fP 'u
[
.B \-no-user-file
]
.if t .ti +\w'\fBhoc\fP 'u
[
.B \-quick
]
[
.B \-secure
]
[
.B \-silent
]
.if n .ti +\w'\fBhoc\fP 'u
[
.B \-trace-file-opening
]
[
.BI \-U name
]
[
.B \-version
]
.if n .ti +\w'\fBhoc\fP 'u
[
.I "file .\|.\|."
]
.PP
.B hoc
can be built in up to three floating-point
precisions, corresponding to the C/C++ data types
.IR float ,
.IR double ,
and
.IR "long double" .
These are conventionally distinguished by suffixes
indicating the number of bits in the
floating-point system:
.BR hoc32 ,
.B hoc64
(same as
.BR hoc ),
.BR hoc80 ,
and
.BR hoc128 .
Support for precisions other than 64-bit
.I double
is deficient, or nonexistent, in C/C++
implementations on several operating systems, so
some
.B hoc
precisions may be unavailable on your system.
.\" ----------------------------------------------
.SH OPTIONS
.B hoc
options can be prefixed with either one or two
hyphens, and can be abbreviated to any unique
prefix.  Thus,
.BR \-a ,
.BR \-author ,
and
.B \-\-auth
are equivalent.
.PP
To avoid confusion with options, if a filename
begins with a hyphen, it must be disguised by a
leading absolute or relative directory path, e.g.,
.I /tmp/-foo.hoc
or
.IR ./-foo.hoc .
.\"-----------------------------------------------
.TP \w'\-no-translation-file'u+3n
.B \-author
Display an author credit, and software
distribution information, on the standard error
unit,
.IR stderr ,
and then terminate with a success return code.
Sometimes an executable program is separated from
its documentation and source code; this option
provides a way to recover from that.
.\"-----------------------------------------------
.TP
.B \-copyright
Display copyright information on the standard
error unit,
.IR stderr ,
and then terminate with a success return code.
.\"-----------------------------------------------
.TP
.BI \-D name
Define the numeric variable
.I name
to have the value 1.
.\"-----------------------------------------------
.TP
.BI \-D name=number
Define the numeric variable
.I name
to have the value
.IR number .
.IP
If = is changed to :=, the assignment is
permanent:
.I name
cannot then subsequently be redefined.
.\"-----------------------------------------------
.TP
.BI \-D name="string"
Define the string variable
.I name
to have the value
.IR \&"string" .
.IP
If = is changed to :=, the assignment is
permanent:
.I name
cannot then subsequently be redefined.
.IP
Since command shells on some operating systems
interpret quotation marks, it is usually necessary
to protect them.  On UNIX-like systems, use
.BI '\-D name="string" '
or if the value contains no characters that are
significant to the command shell,
.BI \-D name=\e"string\e"
.\"-----------------------------------------------
.TP
.BI \-D name=symbol
Define the numeric or string variable
.I name
to have the value of that of
.IR symbol ,
which must be an existing named constant or
variable.
.IP
If = is changed to :=, the assignment is
permanent:
.I name
cannot then subsequently be redefined.
.\"-----------------------------------------------
.TP
.BR \-help " or " \-?
Display a help message on
.IR stderr ,
giving a brief usage description, and then
terminate with a success return code.
.\"-----------------------------------------------
.TP
.B \-no-banner
Suppress any welcome banners normally printed by
dynamically-loaded library code.
.IP
This option can also be set via the
.B hoc
system variable
.BR _\|_BANNER_\|_ ,
but it must be set in an initialization file
before code in that file to print the welcome
banner is reached.
.\"-----------------------------------------------
.TP
.B \-no-cd
Disable the change-directory function,
.BI cd( dir ) ,
and the print-working-directory function,
.BR pwd(\|) .
.IP
This option is a security feature: it takes effect
only
.I after
all initialization files have been processed.
.\"-----------------------------------------------
.TP
.B \-no-help-file
Suppress loading of system-wide
.B hoc
help files at startup.
.\"-----------------------------------------------
.TP
.B \-no-load
Disable the
.B load(\|)
function.  It will continue to be recognized, but
when invoked, will simply print a warning that it
has been disabled.
.IP
This option is a security feature: it takes effect
only
.I after
all initialization files have been processed.
.\"-----------------------------------------------
.TP
.B \-no-logfile
Disable the
.BR logfile(\|) ,
.BR logon(\|) ,
and
.B logoff(\|)
functions.  They will continue to be recognized,
but when invoked, will simply print a warning that
they have been disabled.
.IP
This option is a security feature: it takes effect
only
.I after
all initialization files have been processed.
.\"-----------------------------------------------
.TP
.B \-no-readline
Suppress use of the GNU
.I readline
library: command completion, editing and recall
are then not available.
.IP
On some systems, it may be necessary to use this
option when
.B hoc
is used in international mode (see the
.B INTERNATIONALIZATION
section below) in order to get accented letters
displayed properly.
.\"-----------------------------------------------
.TP
.B \-no-save
Disable the
.B save(\|)
function.  It will continue to be recognized, but
when invoked, will simply print a warning that it
has been disabled.
.IP
This option is a security feature: it takes effect
only
.I after
all initialization files have been processed.
.\"-----------------------------------------------
.TP
.B \-no-site-file
Suppress loading of the system-wide non-help
startup files.
.\"-----------------------------------------------
.TP
.B \-no-translation-file
Suppress loading of the system-wide message
translation files.
.\"-----------------------------------------------
.TP
.B \-no-user-file
Suppress loading of the user-specific startup
file.
.\"-----------------------------------------------
.TP
.B \-quick
Suppress loading of all startup files: this option
is equivalent to
.B \-no-help-file
.B \-no-site-file
.B \-no-translation-file
.BR \-no-user-file .
.\"-----------------------------------------------
.TP
.B \-secure
Enable all security features:  this option is
equivalent to
.B \-no-cd
.B \-no-load
.B \-no-logfile
.BR \-no-save ,
and in addition, makes it impossible to trace file
openings by setting the
.B _\|_DEBUG_OPEN_\|_
system variable to a nonzero value.
.\"-----------------------------------------------
.TP
.B \-silent
Suppress printing of prompts for interactive
input.
.B hoc
never prompts when it is reading noninteractive
files.
.IP
The
.B hoc
system variable
.B _\|_VERBOSE_\|_
can also be set to zero at run time to turn off
prompts; setting it to nonzero turns them back on.
.IP
The
.B hoc
system variable
.B _\|_PROMPT_\|_
contains the prompt string: it can be redefined at
any time.
.\"-----------------------------------------------
.TP
.BI \-trace-file-opening
Set the
.B hoc
variable
.B _\|_DEBUG_OPEN_\|_
to a nonzero value to trace input file opening
attempts.
.\"-----------------------------------------------
.TP
.BI \-U name
Undefine the variable
.IR name .
.\"-----------------------------------------------
.TP
.B \-version
Display the program version number and release
date on
.IR stderr ,
and then terminate with a success return code.
.\" ----------------------------------------------
.SH DESCRIPTION
.B hoc
interprets a simple language for floating-point
arithmetic, at about the level of Basic, with
C-like syntax and functions.
However, unlike Basic,
.B hoc
has particularly rich support for floating-point
arithmetic, and its facilities are certainly
better than that standardly provided by most
programming languages, such as C, C++, and
Fortran.
.PP
.B hoc
recognizes the three popular line-ending
conventions in text files: line feed (LF) (UNIX),
carriage return (CR) (Apple MacOS), and CR LF (PC
DOS, Microsoft Windows, and several older
systems).  It also ignores one or more Ctl-Z
characters at end of file, a horrid relic of some
legacy desktop operating systems.  Thus, its input
files should not even require line-terminator
translation when they are moved between systems.
.PP
To get a flavor of what typical
.B hoc
code looks like, visit the \&\fC*.hoc\fP and
\&\fC*.rc\fP files in the
.B hoc
installation directory tree: see the
.B "INITIALIZATION FILES"
section below for their location.
.PP
The named
.IR file s
are read and interpreted in order.
If no
.I file
is given or if
.I file
is
.IR \- ,
.B hoc
interprets the standard input.
.PP
See the
.B "INPUT FILE SEARCH PATH"
section below for details on how
.B hoc
finds input files.
.PP
.B hoc
input consists of
.I expressions
and
.IR statements .
Expressions are evaluated and their results
printed.  Statements, typically assignments and
function or procedure definitions, produce no
output unless they explicitly call
.BR print .
.\" ..............................................
.SS "Word completion"
When
.B hoc
has been built with the GNU
.I readline
library, word completion can be used to save
typing effort.  It is normally requested by an
ESCape character following a prefix of a word in
.BR hoc 's
symbol table:
.B hoc
will respond with an audible beep, and a list of
words that match that prefix, or if only one word
matches, it will silently complete the word:
.RS
.nf
\&\fC\fBhoc>\fP c<ESCape>
cbrt ceil copysign cos cosd cosh
\fBhoc>\fP co<ESCape>
copysign cos cosd cosh
\fBhoc>\fP cop<ESCape>
\fBhoc>\fP copysign\fP
.fi
.RE
.\" groff bug: bold mode continues into next line,
.\" so explicitly reset the font:
\&\fRThe character used to request completion can
be changed: see the\fP
.B "INITIALIZATION FILES"
section below.
.\" ..............................................
.SS "Command history and editing"
When
.B hoc
has been built with the GNU
.I readline
library, convenient command history and editing
support is available, much like it is in the
UNIX
.BR bash (1),
.BR ksh (1)
and
.BR tcsh (1)
shells, and in a few GNU programs, like
the
.BR bc (1)
and
.BR genius (1)
calculators.   The default history and editing
mode is
.BR emacs (1)-style;
you can also get
.BR vi (1)-style
by suitable customizations: see the
.B "INITIALIZATION FILES"
section below.
.PP
In the default mode,
.I C-p
(hold the Control key down while typing p) moves
up in the history list,
.I C-n
moves down,
.I C-b
moves backward in the current line,
.I C-f
moves forward,
.I C-d
deletes forward,
.I DELete
deletes backward,
.I C-a
moves to the beginning of the line,
.I C-e
moves to the end of the line,
.I C-l
repaints the screen, reprinting the current line
at the top, and
.I RETurn
resubmits the line for execution.
.PP
For more details, consult the
.I "GNU Readline Library"
manual, available online in the
.I info
system.  In
.BR emacs (1);
type
.I "C-h i mreadline"
to get there.
.\" ..............................................
.SS Numbers
All numbers in
.B hoc
are stored as double-precision floating-point
values.
.PP
On systems with IEEE 754 arithmetic, such numbers
are capable of representing integers of up to 53
bits exactly, excluding the sign bit.  This is an
integer range of \-(2^53) .\|.\|. 2^53, or
\-9,007,199,254,740,992
\&.\|.\|. 9,007,199,254,740,992.
.PP
Some systems may have
.B hoc
variants named
.BR hoc32 ,
.BR hoc64 ,
.BR hoc80 ,
or
.BR hoc128 ,
in which case,
.BR hoc 's
default precision is indicated by the program
name: 32-bit, 64-bit, 80-bit, or 128-bit.
.B hoc
and
.B hoc64
are identical on all systems with IEEE 754
arithmetic.
.PP
For
.BR hoc32 ,
on systems with IEEE 754 arithmetic, numbers are
capable of representing integers up to 24 bits
exactly, excluding the sign bit.  This is an
integer range of \-(2^24) \&.\|.\|. 2^24, or
\-16,777,216 \&.\|.\|. +16,777,216.
.PP
For
.BR hoc80 ,
on systems with IEEE 754 arithmetic, numbers are
capable of representing integers up to 64 bits
exactly, excluding the sign bit.  This is an
integer range of \-(2^64) \&.\|.\|. 2^64, or
\-18,446,744,073,709,551,616
\&.\|.\|. +18,446,744,073,709,551,616.
.PP
For
.BR hoc128 ,
on systems with IEEE 754 arithmetic, numbers are
capable of representing integers up to 112 bits
exactly, excluding the sign bit.  This is an
integer range of \-(2^112) \&.\|.\|. 2^112, or
\-5,192,296,858,534,827,628,530,496,329,220,096
\&.\|.\|.
+5,192,296,858,534,827,628,530,496,329,220,096.
.PP
Numbers may be signed, and may optionally contain
a decimal point, a power-of-ten exponent, and a
precision suffix.  The exponent consists of an
exponent letter, one of
.IR d ,
.IR D ,
.IR e ,
.IR E ,
.IR q ,
or
.IR Q
(supported by one or more of Ada, C, C++, Fortran,
Java, and Pascal), followed by an
optionally-signed integer.  The precision suffix
(used by C, C++, and Java) is one of
.IR f ,
.IR F ,
.IR l ,
or
.IR L .
The suffix does
.I not
affect the precision of the constant for
.BR hoc :
it is recognized only to simplify incorporation
into
.B hoc
programs of numbers from programs in other
languages, and their output.
.PP
A hexadecimal floating-point number format,
introduced in the latest ISO C Standard,
.IR "ISO/IEC 9899:1999 (E) Programming languages \(em C" ,
usually known by its short name,
.IR C99 ,
is also supported, and implemented by portable
private code in
.BR hoc .
This format consists of an optional sign, then
.I 0x
or
.IR 0X ,
followed by one or more hexadecimal digits
.RI ( "0\(mi9 A\(miF a\(mif" )
containing at most one hexadecimal point,
optionally followed by a binary (power-of-two)
exponent consisting of
.I p
or
.I P
followed by an optionally-signed decimal integer.
If present, the exponent may optionally be
followed by a precision suffix letter: one of
.IR f ,
.IR F ,
.IR l ,
or
.IR L .
Thus,
.IR \-0x1.00000p8 ,
.IR \-0x100 ,
.IR \-0x100000p-12f ,
.IR \-0x10p+4L ,
.IR \-0x1p+8 ,
.IR \-0x1p00008 ,
and
.I \-0x1p8
all represent the decimal number
.IR \-256 .
.PP
The hexadecimal format, while awkward for humans,
has the advantage of guaranteeing exact
input/output conversions on all platforms, and
.B hoc
consequently uses this format in files created by
the
.B save(\|)
command.
.\" ..............................................
.SS Strings
String constants are delimited by quotation marks
(\fC".\|.\|."\fP), and may not span multiple
lines, unless the embedded line breaks are each
prefixed with a backslash, which is removed,
leaving the newline in the string.
.PP
All characters in 1 .\|.\|. 255 are representable
in strings; as in C and C++, character 0 (ASCII
NUL) is reserved as a string terminator.
.PP
In string constants, nonprintable characters may
be represented by the usual escape sequences
defined in Standard C and Standard C++, plus one
extension (\fB\eE\fP):
.RS
.TP \w'\fB\eo\ \eoo\ \eooo\fP'u+2n
.B \e\e
backslash: ASCII decimal 92.
.TP
.B \e"
quotation mark: ASCII decimal 34.
.TP
.B \ea
alert or bell (ASCII BEL: decimal 7).
.TP
.B \eb
backspace (ASCII BS: decimal 8).
.TP
.B \eE
escape (ASCII ESC: decimal 27).
.TP
.B \ef
formfeed (ASCII FF or NP: decimal 12).
.TP
.B \en
newline (ASCII LF or NL: decimal 10).
.TP
.B \er
carriage return (ASCII CR: decimal 13).
.TP
.B \et
horizontal tab (ASCII HT: decimal 9).
.TP
.B \ev
vertical tab (ASCII VT: decimal 11).
.TP
.B "\eo \eoo \eooo"
octal character number (\fBo\fP = \fI0\(mi7\fP) in
one to three digits.
.TP
.B \exh.\|.\|.
hexadecimal character (\fBh\fP =
\&\fI0\(mi9A\(miF\fP or \fI0\(mi9a\(mif\fP) in one
or more digits.
.RE
Backslash followed by any other character than
those listed is simply discarded: \fB\eW\fP
reduces to \fBW\fP.
.\" ..............................................
.SS Variables
Variable names consist of an initial letter or
underscore, followed by any number of letters,
underscores, or digits.  Lettercase is
.IR significant .
Letters are considered to be
.IR A\(miZ ,
.IR a\(miz ,
and any characters in the range 160 \&.\|.\|. 255
of an 8-bit character set.  Use of characters in
the latter range is normally not recommended,
because they are often difficult, or impossible,
to generate on some computer keyboards.
Nevertheless, it does permit non-English words to
be spelled correctly; see the
.B INTERNATIONALIZATION
section below.
.PP
Underscore
.RI "(\|" _ "\|)"
by itself is a reserved variable containing the
value of the last
.I numeric
expression evaluated.  Double underscore
.RI "(\|" _\|_ "\|)"
is a reserved variable containing the value of the
last
.I string
expression evaluated.  They cannot be assigned to
by user code.
.\" ..............................................
.SS "Predefined numeric constants and variables"
Certain immutable named constants are already
initialized:
.RS
.\"-----------------------------------------------
.TP \w'\fBINF\fP\ or\ \fBInf\fP\ or\ \fBInfinity\fP'u+3n
.B CATALAN
.na
Catalan's constant:
sum((\-1)^i/(2*i+1)^2, i = 0..infinity) =
.if n .br
approximately 0.915965594177219015054603514932.\|.\|.
.ad
.\"-----------------------------------------------
.TP
.B CLASS_xxx
One of eleven possible return values from the
.B class(x)
function; see the section
.B "DESCRIPTIONS OF BUILT-IN FUNCTIONS AND PROCEDURES"
below for details.
.\"-----------------------------------------------
.TP
.B DEG
180/PI, degrees per radian
.\"-----------------------------------------------
.TP
.B E
base of natural logarithms
.\"-----------------------------------------------
.TP
.B GAMMA
Euler's constant:
.br
.na
limit(sum(1/i,i=1.\|.\|.n) \(mi\& ln(n), n \(-> infinity) =
approximately 0.577215664901532860606512090082.\|.\|.
.ad
.\"-----------------------------------------------
.TP
.BR INF " or " Inf " or " Infinity
IEEE-754 floating-point infinity
.\"-----------------------------------------------
.TP
.B MAXNORMAL
Largest finite normalized floating-point number.
.\"-----------------------------------------------
.TP
.B MINNORMAL
Smallest (in absolute value) nonzero normalized
floating-point number.
.\"-----------------------------------------------
.TP
.B MINSUBNORMAL
Smallest (in absolute value) subnormal
floating-point number.  If your computer system
does not support subnormal numbers, this is
identical to
.BR MINNORMAL .
.\"-----------------------------------------------
.TP
.BR NAN " or " NaN
IEEE-754 floating-point not-a-number
.\"-----------------------------------------------
.TP
.B PHI
.na
golden ratio:
.if n .br
(1 + sqrt(5))/2 =
.if n .br
approximately 1.61803398874989484820458683436.\|.\|.
.ad
.\"-----------------------------------------------
.TP
.B PI
.na
ratio of the circumference of a circle to its
diameter, approximately
3.14159265358979323846264338327.\|.\|.
.ad
.\"-----------------------------------------------
.TP
.B PREC
maximum number of significant digits in output,
initially 17 on most systems (the precise value is
computed dynamically, from Matula's 1968 result:
.IR "ceil(N/log_b(10) + 1)" ,
for a host floating-point system with
.I N
.RI base- b
digits).
.B "PREC = 0"
gives shortest `exact' values.
.\"-----------------------------------------------
.TP
.BR QNAN " or " QNaN
IEEE-754 floating-point quiet not-a-number
.\"-----------------------------------------------
.TP
.BR SNAN " or " SNaN
IEEE-754 floating-point signaling not-a-number
.\"-----------------------------------------------
.RE
.PP
More information on the floating-point constants
is available in the
.B "FLOATING-POINT ARITHMETIC"
section below.
.\" ..............................................
.SS "Predefined system constants and variables"
.B hoc
also provides a number of system constants and
variables, adopting the C/C++ convention that
names beginning with two underscores are reserved
for the implementation:
.RS
.\"-----------------------------------------------
.TP \w'_\|_PACKAGE_BUGREPORT_\|_'u+3n
.B "_"
[immutable number] Value of the last numeric
expression printed (initialized to 0.0 on
startup).
.\"-----------------------------------------------
.TP
.B "_\|_"
[immutable string] Value of the last string
expression printed (initialized to an empty string
on startup).
.\"-----------------------------------------------
.TP
.B _\|_BANNER_\|_
[reassignable number] Nonzero (true) if printing
of welcome banners is permitted.  It can be
changed by the
.B \-no-banner
option.
.\"-----------------------------------------------
.TP
.B _\|_CPU_LIMIT_\|_
[immutable number] Current limit on CPU use, in
seconds.  It is normally infinite, but can be
reset by the
.B cpulimit(\|)
function.
.\"-----------------------------------------------
.TP
.B _\|_DATE_\|_
[constant string] Date of the start of job
execution, in the form \&\fC"Dec 16 2001"\fP.  The
day number has a leading space if only one digit
is needed, so that the string always has constant
width.
.\"-----------------------------------------------
.TP
.B _\|_FILE_\|_
[constant string]
Name of the current input file.  This is
\&\fC"/dev/stdin"\fP when
.B hoc
is reading from the standard input.
.\"-----------------------------------------------
.TP
.BI _\|_FILE_\|_[ n ]
[constant string]
Name of the
.IR n -th
input file in the current job.  This provides a

at/changelog
------------------------------------------------------------------------
r40099 | robot | 2014-07-16 09:30:15 +0200 (Wed, 16 Jul 2014) | 2 lines

UHU 3 (LTS)

------------------------------------------------------------------------
r38755 | janusz@uhulinux.hu | 2014-04-29 13:05:53 +0200 (Tue, 29 Apr 2014) | 2 lines

systemd initscript from fedora

------------------------------------------------------------------------
r36982 | peritus@freemail.hu | 2014-02-27 11:05:03 +0100 (Thu, 27 Feb 2014) | 2 lines

3.1.13 -> 3.1.14

------------------------------------------------------------------------
r33785 | janusz@uhulinux.hu | 2013-09-30 10:32:34 +0200 (Mon, 30 Sep 2013) | 2 lines

doesn't support multithread build

------------------------------------------------------------------------
r33488 | peritus@freemail.hu | 2013-08-31 20:32:10 +0200 (Sat, 31 Aug 2013) | 2 lines

add u2d

------------------------------------------------------------------------
r25911 | peritus@freemail.hu | 2011-08-02 20:52:53 +0200 (Tue, 02 Aug 2011) | 2 lines

3.1.12 -> 3.1.13

------------------------------------------------------------------------
r21434 | janusz@uhulinux.hu | 2009-12-03 18:51:38 +0100 (Thu, 03 Dec 2009) | 2 lines

3.1.11 -> 3.1.12

------------------------------------------------------------------------
r20656 | peritus@freemail.hu | 2009-10-28 15:39:27 +0100 (Wed, 28 Oct 2009) | 3 lines

3.1.8 -> 3.1.11
acquire fix

------------------------------------------------------------------------
r13196 | pozsy | 2006-07-19 18:37:28 +0200 (Wed, 19 Jul 2006) | 2 lines

- update initscript

------------------------------------------------------------------------
r12443 | pozsy | 2006-06-14 12:52:13 +0200 (Wed, 14 Jun 2006) | 2 lines

- update initscript

------------------------------------------------------------------------
r12058 | egmont | 2006-05-22 16:17:36 +0200 (Mon, 22 May 2006) | 2 lines

- init script fix

------------------------------------------------------------------------
r11961 | egmont | 2006-05-09 20:03:47 +0200 (Tue, 09 May 2006) | 2 lines

- gettextize init script

------------------------------------------------------------------------
r9152 | egmont | 2005-08-31 15:54:25 +0200 (Wed, 31 Aug 2005) | 2 lines

- Convert package source to UTF-8

------------------------------------------------------------------------
r6251 | egmont | 2004-06-02 12:24:53 +0200 (Wed, 02 Jun 2004) | 3 lines

- acquire fix
- cleanup

------------------------------------------------------------------------
r5350 | egmont | 2004-01-13 20:14:05 +0100 (Tue, 13 Jan 2004) | 2 lines

priority: important -> standard

------------------------------------------------------------------------
r4015 | egmont | 2003-09-30 13:09:58 +0200 (Tue, 30 Sep 2003) | 2 lines

New section

------------------------------------------------------------------------
r3070 | egmont | 2003-08-29 19:53:35 +0200 (Fri, 29 Aug 2003) | 2 lines

új típusú init szkript

------------------------------------------------------------------------
r2838 | egmont | 2003-08-26 16:08:10 +0200 (Tue, 26 Aug 2003) | 2 lines

priority fix

------------------------------------------------------------------------
r2718 | egmont | 2003-08-18 18:18:09 +0200 (Mon, 18 Aug 2003) | 2 lines

service fájl

------------------------------------------------------------------------
r133 | pozsy | 2003-02-18 13:21:28 +0100 (Tue, 18 Feb 2003) | 2 lines

1.0-rc2.1 branched.

------------------------------------------------------------------------
r2 | pozsy | 2003-02-14 10:26:24 +0100 (Fri, 14 Feb 2003) | 1 line

1.0-rc2.1
------------------------------------------------------------------------
r1 | root | 2003-02-14 10:10:29 +0100 (Fri, 14 Feb 2003) | 1 line

1.0-rc2
------------------------------------------------------------------------


hoc/addons/usr/share/hoc/hoc.el
;;; -*-emacs-lisp-*-
;;; ====================================================================
;;;  @Emacs-Lisp-file{
;;;     author          = "Nelson H. F. Beebe",
;;;     version         = "1.01",
;;;     date            = "11 January 2002",
;;;     time            = "06:21:40 MST",
;;;     filename        = "hoc.el",
;;;     address         = "Center for Scientific Computing
;;;                        University of Utah
;;;                        Department of Mathematics, 322 INSCC
;;;                        155 S 1400 E RM 233
;;;                        Salt Lake City, UT 84112-0090
;;;                        USA",
;;;     telephone       = "+1 801 581 5254",
;;;     FAX             = "+1 801 585 1640, +1 801 581 4148",
;;;     URL             = "http://www.math.utah.edu/~beebe",
;;;     checksum        = "23163 206 801 7139",
;;;     email           = "beebe@math.utah.edu, beebe@acm.org,
;;;                        beebe@computer.org, beebe@ieee.org
;;;                        (Internet)",
;;;     codetable       = "ISO/ASCII",
;;;     keywords        = "hoc (high-order calculator)",
;;;     supported       = "yes",
;;;     docstring       = "This is an Emacs Lisp function that augments
;;;                        c-mode to make a simple hoc-mode.
;;;
;;;                        The checksum field above contains a CRC-16
;;;                        checksum as the first value, followed by the
;;;                        equivalent of the standard UNIX wc (word
;;;                        count) utility output of lines, words, and
;;;                        characters.  This is produced by Robert
;;;                        Solovay's checksum utility.",
;;;  }
;;; ====================================================================

;;; hoc.el --- hoc (high-order calculator) editing support

;; Author: Nelson H. F. Beebe <beebe@math.utah.edu>
;; Created: 08-Dec-2001
;; Version: 1.00
;; Keywords: hoc (high-order calculator)

;; Copyright (C) 2001--2002 Free Software Foundation, Inc.

;; This file is part of GNU Emacs.

;; GNU Emacs is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation; either version 2, or (at your option)
;; any later version.

;; GNU Emacs is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with GNU Emacs; see the file COPYING.  If not, write to
;; the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.

;;; Commentary:

;;; Change log:

;; ---------------------------------------------------------------------
;; [1.01] Fri Jan 11 06:17:49 2002
;; Add new functions hoc-printify and hoc-unprintify to help in
;; creating the help_xxx() documentation functions.
;;
;; Add provide statement so that this library can be loaded with
;; (require 'hoc).
;;
;; Add this GNU conventional comment header and copyright statement.
;;
;; ---------------------------------------------------------------------
;; [1.00] Sat Dec 8 14:32:29 MST 2001
;; Original version with function hoc-mode.
;; ---------------------------------------------------------------------

(provide 'hoc)

(defun hoc-mode ()
  "Set up for editing hoc code (a minor variation on C mode)."
  (interactive)
  (let ((empty-buffer (= (point-min) (point-max))))
    (goto-char (point-min)) ;prevent c-mode's insertion of C template file
    (if empty-buffer
        (insert "#!\n"))
    (c-mode)
    (make-local-variable 'comment-start)
    (setq comment-start "# ")
    (make-local-variable 'comment-end)
    (setq comment-end "")
    (make-local-variable 'comment-column)
    (make-local-variable 'comment-start-skip)
    (setq comment-start-skip "# *")
    (setq mode-name "hoc")
    (setq c-argdecl-indent 0
	  c-auto-newline t
	  c-basic-offset 8
	  c-brace-imaginary-offset 0
	  c-brace-offset 0
	  c-continued-statement-offset 0
	  c-indent-level 8
	  c-label-offset -8
	  case-fold-search nil	      ;case sensitive searching in hoc
	  comment-column 40)
    (goto-char (point-min)) ;prevent c-mode's possible insertion of C template file
    (if (and empty-buffer (looking-at "^#!\n"))
	(delete-char 3))
    (run-hooks 'hoc-mode-hook)
    (goto-char (point-min))
    (if empty-buffer
	(progn
	  (insert "### -*-hoc-*-\n")
	  (insert "### ====================================================================\n")
	  (insert "### \n")
	  (insert "### ")
	  (date)
	  (insert "\n")
	  (insert "### ====================================================================\n")
	  (goto-char (point-min))
	  (forward-line 2)
	  (forward-char 4)
	  (set-buffer-modified-p nil)))))

(if (string-lessp (substring emacs-version 0 2) "19") ; earlier than 19.x
    (progn
      (setq auto-mode-alist
	    (cons (cons "\\.hoc$" 'hoc-mode) auto-mode-alist))
      (autoload 'hoc-mode  "hoc"
	"Enter hoc mode." t nil))
  (progn
    (if (not (assoc "\\.hoc$" auto-mode-alist))
	(setq auto-mode-alist
	      (cons (cons "\\.hoc$" 'hoc-mode) auto-mode-alist)))
    (autoload 'hoc-mode  "hoc"
      "Enter hoc (high-order calculator) mode." t nil)))

(defun hoc-printify()
  "Convert the region, assumed to be formatted to the correct line
width, with the correct number of leading tabs, to a series of hoc
print statements.

This function should normally be applied to a single paragraph of
documentation."
  (interactive)
  (save-excursion
    (save-restriction
      (narrow-to-region (mark) (point))
      (goto-char (point-min))
      (while (search-forward "\"" nil t) ;protect embedded quotes
	(replace-match "\\\"" nil t))
      (goto-char (point-min))
      (while (< (point) (point-max))
	(progn
	  (insert "\tprint \"")
	  (while (looking-at "\t")
	    (progn
	      (delete-char 1)
	      (insert "\\t")))
	  (end-of-line)
	  (delete-horizontal-space)
	  (if (>= (1+ (point)) (point-max))
	      (insert "\\n"))
	  (insert "\\n\"")
	  (forward-line 1))))))


(defun hoc-unprintify()
  "Convert the region, assumed to consist of a series of hoc print
statements, to the text it would print if executed.

This function should normally be applied to a single paragraph of
documentation."
  (interactive)
  (save-excursion
    (save-restriction
      (narrow-to-region (mark) (point))
      (goto-char (point-min))
      (while (< (point) (point-max))
	(progn
	  (if (looking-at "\tprint \"")
	      (progn
		(delete-char 8)
		(while (looking-at "\\\\t")
		  (progn
		    (delete-char 2)
		    (insert "\t")))
		(end-of-line)
		(delete-horizontal-space)
		(backward-char 5)
		(if (looking-at "\\\\n\\\\n\"")
		    (progn
		      (delete-char 5)
		      (insert "\n"))
		  (forward-char 2)
		  (if (looking-at "\\\\n\"")
		      (delete-char 3)))
		(forward-line 1)))))
      (goto-char (point-min))
      (while (search-forward "\\\"" nil t) ;reduce protected quotes
	(replace-match "\"" nil t))
      (goto-char (point-max))
      (delete-blank-lines))))
